src/librustdoc/html/render/print_item.rs

```diff
@@ -1936,111 +1938,77 @@ fn document_type_layout<'a, 'cx: 'a>(
-                    <strong>completely unstable</strong> and may even differ between compilations. \
-                    The only exception is types with certain <code>repr(...)</code> attributes. \
-                    Please see the Rust Reference’s \
-                    <a href=\"https://doc.rust-lang.org/reference/type-layout.html\">“Type Layout”</a> \
-                    chapter for details on type layout guarantees.</p></div>"
-                )?;
-                f.write_str("<p><strong>Size:</strong> ")?;
-                write_size_of_layout(&mut f, &ty_layout.layout.0, 0);
-                writeln!(f, "</p>")?;
-                if let Variants::Multiple { variants, tag, tag_encoding, .. } =
-                    &ty_layout.layout.variants()
-                {
-                    if !variants.is_empty() {
-                        f.write_str(
-                            "<p><strong>Size for each variant:</strong></p>\
-                                <ul>",
+        fn type_layout<'b: 'a>(&'b self) -> Result<TyAndLayout<'cx>, LayoutError<'cx>> {
+            let tcx = self.cx.tcx();
+            let param_env = tcx.param_env(self.ty_def_id);
+            let ty = tcx.type_of(self.ty_def_id).subst_identity();
+            tcx.layout_of(param_env.and(ty))
+        }
+        fn variant_name<'b: 'a>(&'b self, index: VariantIdx) -> Symbol {
+            let Adt(adt, _) = self.type_layout().unwrap().ty.kind() else {
+                span_bug!(self.cx.tcx().def_span(self.ty_def_id), "not an adt")
+            };
+            adt.variant(index).name
+        }
+        fn tag_size<'b: 'a>(&'b self) -> u64 {
+            if let Variants::Multiple { variants, tag, tag_encoding, .. } =
+                    self.type_layout().unwrap().layout.variants() && !variants.is_empty() {
+                if let TagEncoding::Niche { .. } = tag_encoding {
+                    0
+                } else if let Primitive::Int(i, _) = tag.primitive() {
+                    i.size().bytes()
+                } else {
+                    span_bug!(self.cx.tcx().def_span(self.ty_def_id), "tag is neither niche nor int")
+                }
+            } else {
+                0
+            }
+        }
+        fn write_size<'b: 'a>(
+            &'b self,
+            layout: &'b LayoutS,
+            tag_size: u64,
+        ) -> impl fmt::Display + Captures<'cx> + Captures<'b> {
+            display_fn(move |f| {
+                if layout.abi.is_unsized() {
+                    write!(f, "(unsized)")?;
+                } else {
+                    let size = layout.size.bytes() - tag_size;
+                    write!(f, "{size} byte{pl}", pl = if size == 1 { "" } else { "s" })?;
+                    if layout.abi.is_uninhabited() {
+                        write!(
+                            f,
+                            " (<a href=\"https://doc.rust-lang.org/stable/reference/glossary.html#uninhabited\">uninhabited</a>)"
```

Comment:
This function looks like a pretty good candidate for returning a struct that can be rendered as a Template, e.g.:

```
struct Size {
  is_unsized: bool,
  is_uninhabited: bool,
  size: usize,
}
```

```
{% if is_unsized %}
  (unsized)
{% else %}
  {% if size == 1 %}
    1 byte
  {% else %}
    {{ size }} bytes
  {% endif %}
  {% if is_uninhabited %}
    (<a href=\"https://doc.rust-lang.org/stable/reference/glossary.html#uninhabited\">uninhabited</a>)
  {% endif %}
{% endif %]
```